tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/tests/scoping/ComposedMemberCachingTest.xtend

```diff
@@ -0,0 +1,109 @@
+/**
+ * Copyright (c) 2017 NumberFour AG.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   NumberFour AG - Initial API and implementation
+ */
+package org.eclipse.n4js.tests.scoping
+
+import com.google.inject.Inject
+import org.eclipse.n4js.AbstractN4JSTest
+import org.eclipse.n4js.n4JS.MemberAccess
+import org.eclipse.n4js.n4JS.ParameterizedPropertyAccessExpression
+import org.eclipse.n4js.n4JS.VariableDeclaration
+import org.eclipse.n4js.resource.N4JSResource
+import org.eclipse.n4js.scoping.members.MemberScopingHelper
+import org.eclipse.n4js.ts.typeRefs.TypeRef
+import org.eclipse.n4js.ts.types.TMember
+import org.eclipse.n4js.ts.utils.TypeUtils
+import org.eclipse.xtext.naming.IQualifiedNameConverter
+import org.junit.Test
+
+
+/**
+ * Tests behavior of member scoping related to the caching of composed members. For further information, see
+ * API documentation of methods:
+ * {@link MemberScopingHelper#createMemberScope(TypeRef, MemberAccess, boolean, boolean)} and
+ * {@link MemberScopingHelper#createMemberScopeAllowingNonContainedMembers(TypeRef, EObject, boolean, boolean)}.
+ */
+class ComposedMemberCachingTest extends AbstractN4JSTest {
+
+	@Inject
+	private MemberScopingHelper memberScopingHelper;
+	@Inject
+	private IQualifiedNameConverter qualifiedNameConverter;
+
+
+	val snippet = '''
+			let a: Array<string> | Array<number>;
+			a.length;
+	''';
+
+
+	@Test
+	def void testContainedMemberWhenUsingContainedTypeRef() {
+		val script = snippet.parse; // do not validate!
+		val resource = script.eResource as N4JSResource;
+
+		val varDecl = script.eAllContents.filter(VariableDeclaration).head;
+		val propAccess = script.eAllContents.filter(ParameterizedPropertyAccessExpression).head;
+
+		// use a receiver type reference that is contained in a resource
+		val receiverTypeRef = varDecl.declaredTypeRef;
+		val member = performOrdinaryMemberScoping(receiverTypeRef, propAccess);
+		assertTrue("composed member should be contained in a resource", member.eResource!==null);
+		assertSame(resource, member.eResource);
+	}
+
+	@Test
+	def void testContainedMemberWhenUsingNonContainedTypeRef() {
+		val script = snippet.parse; // do not validate!
+		val resource = script.eResource as N4JSResource;
+
+		val varDecl = script.eAllContents.filter(VariableDeclaration).head;
+		val propAccess = script.eAllContents.filter(ParameterizedPropertyAccessExpression).head;
+
+		// use a receiver type reference that is *NOT* contained in a resource
+		val receiverTypeRef = TypeUtils.copy(varDecl.declaredTypeRef);
```

Comment:
well, it won't hurt and I have to make another commit anyway, so I added that assertion

done.